File: common/2021.07.20/amymeme_1644.kt
# https://www.acmicpc.net/problem/1504
# 특정한 최단 경로

import kotlin.math.sqrt

fun getPrimeList(N: Int): List<Int> {
    val sieve = BooleanArray(N + 1) { true }
    for (i in 2..sqrt(N.toDouble()).toInt()) { // N 포함
        if (!sieve[i]) continue
        for (j in i + i..N step i) sieve[j] = false
    }
    return sieve.withIndex().filter { it.index > 1 && it.value }.map { it.index }
}

fun solution(N: Int): Int {
    var count = 0
    val primeList = getPrimeList(N)
    val len = primeList.size

    if (len == 0) return 0

    var left = 0
    var localSum = primeList[left]
    var right = 0

    while (left < len) {
        while (right < len - 1 && localSum + primeList[right + 1] <= N) {
            right += 1
            localSum += primeList[right]
        }
        if (localSum == N) {
            count += 1
        }
        localSum -= primeList[left]
        left += 1
    }
    return count
}


fun main() {
    val n: Int = readLine()!!.toInt()
    println(solution(n))
}

File: common/2021.07.20/amymeme_1644.py
# https://www.acmicpc.net/problem/1644
# 소수의 연속합

import sys
import math

N = int(sys.stdin.readline())
if N == 1:
    print(0)
    exit(0)
prime_list = []
for i in range(2, N + 1):
    is_prime = True
    for j in range(2, int(math.sqrt(i)) + 1):
        if i % j == 0:
            is_prime = False
            break
    if is_prime:
        prime_list.append(i)

# initialize
left = 0
local_sum = prime_list[0]
right = 0
length = len(prime_list)
match_count = 0

while left < length:
    while right < length - 1:
        if local_sum + prime_list[right + 1] > N:
            break
        right += 1
        local_sum += prime_list[right]
    if local_sum == N:
        match_count += 1
    local_sum -= prime_list[left]
    left += 1
print(match_count)
